p2p/src/cybershake.rs

```diff
@@ -218,73 +236,246 @@ where
     Ok((received_remote_identity, outgoing, incoming))
 }
 
-// TODO: implement AsyncWrite for this, buffering the data and encrypting on flush or on each N-byte chunk.
+macro_rules! ready {
+    ($($tokens:tt)*) => {
+        match $($tokens)* {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(Ok(n)) => { n }
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+        }
+    };
+}
+
 impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    /// Send a message with any length.
     pub async fn send_message(&mut self, msg: &[u8]) -> Result<(), Error> {
+        let mut written = 0;
+        while written != msg.len() {
+            match self.write(&msg[written..]).await {
+                Ok(n) => written += n,
+                Err(e) => {
+                    return match e.kind() {
+                        io::ErrorKind::InvalidData => Err(Error::ProtocolError),
+                        _ => Err(Error::IoError(e)),
+                    }
+                }
+            };
+        }
+        self.flush().await.map_err(|e| Error::IoError(e))?;
+        Ok(())
+    }
+}
+
+impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    fn cipher_buf(&mut self) {
+        self.flushing = true;
         self.kdf.append_u64(b"seq", self.seq);
         let mut key = [0u8; 32];
         self.kdf.challenge_bytes(b"key", &mut key);
 
         let ad = encode_u64le(self.seq);
 
         let ciphertext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
-            .encrypt(&[&ad], msg)
-            .map_err(|_| Error::ProtocolError)?;
+            .encrypt(&[&ad], &self.buf)
```

Comment:
You can use `encrypt_in_place` to avoid copying, and make sure that the buffer had 16-byte (IV_SIZE) padding before the plaintext data. Also leave 2 bytes for the length prefix.

Reply:
implement